app/protocol/manager.go

```diff
@@ -32,6 +36,14 @@ func NewManager(cfg *config.Config, domain domain.Domain, netAdapter *netadapter
 	return &manager, nil
 }
 
+// Close closes the protocol manager and waits until all p2p flows
+// finish.
+func (m *Manager) Close() {
+	atomic.StoreUint32(&m.isClosed, 1)
```

Comment:
I think you should compare and swap here, and panic if for some reason isClosed is already 1.